File: epi_judge_cpp/search_first_greater_value_in_bst.cc
#include <memory>

#include "bst_node.h"
#include "test_framework/generic_test.h"
using std::unique_ptr;

// class FindFirst {

//   public:
//   FindFirst(const unique_ptr<BstNode<int>>& tree,
//                                     int k){

//                                     }
// }
class FindFirst
{

public:
  BstNode<int> *bstN = nullptr;
  int minGtVal = std::numeric_limits<int>::max();

  void FindFirstGreaterThanK(const unique_ptr<BstNode<int>> &tree,
                             int k)
  {
    // TODO - you fill in here.
    if (tree == nullptr)
    {
      return;
    }

    if (tree->data > k)
    {
      if (tree->data < minGtVal)
      {
        minGtVal = tree->data;
        bstN = tree.get();
      }
      FindFirstGreaterThanK(tree->left, k);
    }
    else
    {
      FindFirstGreaterThanK(tree->right, k);
    }
  }
};

BstNode<int> *FindFirstGreaterThanK(const unique_ptr<BstNode<int>> &tree, int k)
{
  BstNode<int> *subtree = tree.get();
  BstNode<int> *first_so_far = nullptr;
  while (subtree)
  {
    if (subtree->data > k)
    {
      first_so_far = subtree; // if 
      subtree = subtree->left.get();
    }
    else
    {
      subtree = subtree->right.get();
    }
  }
  return first_so_far;
}
int FindFirstGreaterThanKWrapper(const unique_ptr<BstNode<int>> &tree, int k)
{
  auto result = FindFirstGreaterThanK(tree, k);
  return result ? result->data : -1;
}

int main(int argc, char *argv[])
{
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"tree", "k"};
  return GenericTestMain(args, "search_first_greater_value_in_bst.cc",
                         "search_first_greater_value_in_bst.tsv",
                         &FindFirstGreaterThanKWrapper, DefaultComparator{},
                         param_names);
}

File: epi_judge_python/search_first_greater_value_in_bst.py
from typing import Optional

from bst_node import BstNode
from test_framework import generic_test


def find_first_greater_than_k(tree: BstNode, k: int) -> Optional[BstNode]:
    # TODO - you fill in here.
    minval = float("inf")
    minValNode = None

    def aux(tree):
        nonlocal minval
        nonlocal minValNode
        if tree is None:
            return

        if k < tree.data:
            # might be in left
            # if minval > tree.data: this will always be the case because any number we would have seen before this would have to
            # be bigger via the BST property so this is a moot point
            minval = tree.data
            minValNode = tree
            aux(tree.left)
        else:
            aux(tree.right)

    aux(tree)
    return minValNode


def find_first_greater_than_k_wrapper(tree, k):
    result = find_first_greater_than_k(tree, k)
    return result.data if result else -1


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main(
            'search_first_greater_value_in_bst.py',
            'search_first_greater_value_in_bst.tsv',
            find_first_greater_than_k_wrapper))
